parser_advanced_test.go

```diff
@@ -0,0 +1,386 @@
+		Expr ExprPrec3 `@@`
+	}
+
+	MemberBinRem struct {
+		Head ExprOperand        `@@`
+		Tail []MemberBinRemCont `@@+`
+	}
+
+	MemberBinRemCont struct {
+		Op   string      `@"%"`
+		Expr ExprOperand `@@`
+	}
+
+	ExprPrecAll interface{ exprPrecAll() }
+	ExprPrec1   interface{ exprPrec1() }
+	ExprPrec2   interface{ exprPrec2() }
+	ExprPrec3   interface{ exprPrec3() }
+	ExprOperand interface{ exprOperand() }
+)
+
+// These expression types can be matches as individual operands
+func (MemberIdent) exprOperand()  {}
+func (MemberNumber) exprOperand() {}
+func (MemberString) exprOperand() {}
+func (MemberParens) exprOperand() {}
+func (MemberUnary) exprOperand()  {}
+
+// These expression types can be matched at precedence level 3
+func (MemberIdent) exprPrec3()  {}
+func (MemberNumber) exprPrec3() {}
+func (MemberString) exprPrec3() {}
+func (MemberParens) exprPrec3() {}
+func (MemberUnary) exprPrec3()  {}
+func (MemberBinRem) exprPrec3() {}
+
+// These expression types can be matched at precedence level 2
+func (MemberIdent) exprPrec2()     {}
+func (MemberNumber) exprPrec2()    {}
+func (MemberString) exprPrec2()    {}
+func (MemberParens) exprPrec2()    {}
+func (MemberUnary) exprPrec2()     {}
+func (MemberBinRem) exprPrec2()    {}
+func (MemberBinMulDiv) exprPrec2() {}
+
+// These expression types can be matched at precedence level 1
+func (MemberIdent) exprPrec1()     {}
+func (MemberNumber) exprPrec1()    {}
+func (MemberString) exprPrec1()    {}
+func (MemberParens) exprPrec1()    {}
+func (MemberUnary) exprPrec1()     {}
+func (MemberBinRem) exprPrec1()    {}
+func (MemberBinMulDiv) exprPrec1() {}
+func (MemberBinAddSub) exprPrec1() {}
+
+// These are all of the expression types
+func (MemberIdent) exprPrecAll()     {}
+func (MemberNumber) exprPrecAll()    {}
+func (MemberString) exprPrecAll()    {}
+func (MemberParens) exprPrecAll()    {}
+func (MemberUnary) exprPrecAll()     {}
+func (MemberBinRem) exprPrecAll()    {}
+func (MemberBinMulDiv) exprPrecAll() {}
+func (MemberBinAddSub) exprPrecAll() {}
+
+func TestUnionExprParser(t *testing.T) {
+	type Wrapper struct {
+		Expr ExprPrecAll `@@`
+	}
+
+	var (
+		withExprOperand = participle.UseUnion[ExprOperand](
+			MemberUnary{}, MemberIdent{}, MemberNumber{}, MemberString{}, MemberParens{})
+
+		withExprPrec3 = participle.UseUnion[ExprPrec3](
+			MemberBinRem{}, MemberUnary{}, MemberIdent{}, MemberNumber{}, MemberString{}, MemberParens{})
+
+		withExprPrec2 = participle.UseUnion[ExprPrec2](
+			MemberBinMulDiv{}, MemberBinRem{}, MemberUnary{}, MemberIdent{}, MemberNumber{}, MemberString{}, MemberParens{})
+
+		withExprPrec1 = participle.UseUnion[ExprPrec1](
+			MemberBinAddSub{}, MemberBinMulDiv{}, MemberBinRem{}, MemberUnary{}, MemberIdent{}, MemberNumber{}, MemberString{}, MemberParens{})
+
+		withExprPrecAll = participle.UseUnion[ExprPrecAll](
+			MemberBinAddSub{}, MemberBinMulDiv{}, MemberBinRem{}, MemberUnary{}, MemberIdent{}, MemberNumber{}, MemberString{}, MemberParens{})
```

Comment:
Perhaps this goes without saying, but the order of the arguments to the `UseUnion` function is important, since the parser will attempt to match them in the order given, and will return the first match. 

For example - If the arguments are provided in an order that will attempt to match `a` before attempting to match `a+b`, then it will always match the `a` in `a+b` and will never match `a+b` itself. In my opinion this behavior is ok as long as its documented, but wanted to call it out here in case there's a problem with this that I haven't thought about yet